dart/common/AddonManager.h

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2015, Georgia Tech Research Corporation
+ * All rights reserved.
+ *
+ * Author(s): Michael X. Grey <mxgrey@gatech.edu>
+ *
+ * Georgia Tech Graphics Lab and Humanoid Robotics Lab
+ *
+ * Directed by Prof. C. Karen Liu and Prof. Mike Stilman
+ * <karenliu@cc.gatech.edu> <mstilman@cc.gatech.edu>
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_ADDONMANAGER_H_
+#define DART_COMMON_ADDONMANAGER_H_
+
+#include <map>
+#include <typeinfo>
+#include <typeindex>
+
+#include "dart/common/Addon.h"
+
+namespace dart {
+namespace common {
+
+/// AddonManager is a base class that should be virtually inherited by any class
+/// that wants to be able to manage Addons.
+///
+/// The base AddonManager class is completely agnostic to what kind of Addons it
+/// is given. Addons are stored in a std::map, so access to its Addons happens
+/// on average in log(N) time. Most often, a class that accepts Addons will have
+/// certain Addon types that it will need to access frequently, and it would be
+/// beneficial to have constant-time access to those Addon types. To get
+/// constant-time access to specific Addon types, there are FOUR macros that you
+/// should use in your derived class:
```

Comment:
1. I don't entirely understand the efficiency argument here. Are you currently iterating through the two `std::map`'s in parallel while merging them?

2. This makes sense. I forgot that `std::map` had such nice semantics for iterator invalidation.